src/bootloaders/grub2.c

```diff
@@ -0,0 +1,406 @@
+        cbm_writer_append(writer, "echo \"}\"\n");
+
+        cbm_writer_close(writer);
+
+        conf_path = grub2_get_entry_path_for_kernel((BootManager *)manager, kernel);
+
+        /* If our new config matches the old config, just return. */
+        if (file_get_text(conf_path, &old_conf)) {
+                if (streq(old_conf, writer->buffer)) {
+                        return true;
+                }
+        }
+
+        /* Ensure the grub.d directory actually exists (should do..) */
+        grub_dir = string_printf("%s/etc/grub.d", prefix);
+        if (!nc_file_exists(grub_dir) && !nc_mkdir_p(grub_dir, 00755)) {
+                LOG_FATAL("Failed to create grub.d dir: %s [%s]", grub_dir, strerror(errno));
+                return false;
+        }
+
+        if (!file_set_text(conf_path, writer->buffer)) {
+                LOG_FATAL("Failed to create loader entry for: %s [%s]",
+                          kernel->source.path,
+                          strerror(errno));
+                return false;
+        }
+
+        /* Ensure it's executable */
+        if (chmod(conf_path, 00755) != 0) {
+                LOG_FATAL("Failed to mark loader entry as executable: %s [%s]",
+                          kernel->source.path,
+                          strerror(errno));
+                return false;
+        }
+
+        cbm_sync();
+
+        return true;
+}
+
+bool grub2_remove_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        if (!manager || !kernel) {
+                return false;
+        }
+        autofree(char) *conf_path = NULL;
+
+        conf_path = grub2_get_entry_path_for_kernel((BootManager *)manager, kernel);
+        if (nc_file_exists(conf_path) && unlink(conf_path) < 0) {
+                LOG_FATAL("grub2_remove_kernel: Failed to remove %s: %s",
+                          conf_path,
+                          strerror(errno));
+                return false;
+        }
+        return true;
+}
+
+bool grub2_set_default_kernel(const BootManager *manager, const Kernel *default_kernel)
+{
+        if (!manager) {
+                return false;
+        }
+        autofree(char) *vmlinuz_path = NULL;
+        autofree(char) *initrd_path = NULL;
+        autofree(char) *command = NULL;
+        autofree(char) *boot_dir = NULL;
+        autofree(char) *grub_dir = NULL;
+        autofree(char) *vmlinuz_rel = NULL;
+        autofree(char) *initrd_rel = NULL;
+        autofree(char) *boot_rel = NULL;
+        const char *prefix = NULL;
+        int ret;
+
+        prefix = boot_manager_get_prefix((BootManager *)manager);
+        boot_dir = boot_manager_get_boot_dir((BootManager *)manager);
+        vmlinuz_path = string_printf("%s/vmlinuz", prefix);
+        initrd_path = string_printf("%s/initrd.img", prefix);
+
+        /* Always nuke the files *before* running grub-mkconfig to stop duped
+         * entries being created */
+        if (nc_file_exists(vmlinuz_path) && unlink(vmlinuz_path) < 0) {
```

Comment:
So those links are just that, links. They're only useful on a system where we can determine the currently running kernel, which infers a CBM managed system to start with. We provide those so that `sudo update-grub` works from other distros. We remove them at the start to prevent our own GRUB seeing ourselves twice, and then later restore the links for the dual-boot situation.